### PR TITLE
[RISCV][VLOPT] Enable the RISCVVLOptimizer by default

### PR DESCRIPTION
**Pull request by michaelmaitland**
Now that we have testing of all instructions in the isSupportedInstr switch, and better coverage of getOperandInfo, I think it is a good time to enable this by default.

### CODE REVIEW

**Comment by michaelmaitland** [comment]
ping

**Comment by lukel97** [comment]
Have you had a chance to kick the tires with this on llvm-test-suite or SPEC?

**Comment by michaelmaitland** [comment]
> Have you had a chance to kick the tires with this on llvm-test-suite or SPEC?

I had a chance to run this on spec2006/int/train and spec2017/int/train on qemu. Both build and terminate without errors.

**Comment by preames** [comment]
Reading through the code, I spotted one potential correctness issue.  This is a cornercase, but probably still worth fixing.

Imagine you have the following:
%v = VADD_VV ...
%s = VREDSUM w/ %v as scalar source
%dead = VADD_VV %v, %v w/ VL=0

The last instruction is dead - it can be folded to it's passthru.  (In practice, it probably will have been folded, but it's possible something could slip through to here.)  However, when scaning the users of %v, we will decide that the correct VL for %v is 0 (or a register which might be zero), and reduce it below the minimum VL=1 required by the reduction.

To fix this, I believe you need to treat the CommonVL for the scalar operand case as being VL=1.  You could also track a non-zero state instead.  

Other than that, looks good to me.  Once you've fixed this issue, happy to approve.

**Comment by lukel97** [comment]
> > Have you had a chance to kick the tires with this on llvm-test-suite or SPEC?
> 
> I had a chance to run this on spec2006/int/train and spec2017/int/train on qemu. Both build and terminate without errors.

Nice, this also just came to mind but did you run it with the `rvv_ta_all_1s=1` option set? I'm thinking if there were any potential miscompiles then this would probably be needed to catch them

**Comment by michaelmaitland** [comment]
> > > Have you had a chance to kick the tires with this on llvm-test-suite or SPEC?
> > 
> > 
> > I had a chance to run this on spec2006/int/train and spec2017/int/train on qemu. Both build and terminate without errors.
> 
> Nice, this also just came to mind but did you run it with the `rvv_ta_all_1s=1` option set? I'm thinking if there were any potential miscompiles then this would probably be needed to catch them

Yes

**Comment by topperc** [comment]
> Reading through the code, I spotted one potential correctness issue. This is a cornercase, but probably still worth fixing.
> 
> Imagine you have the following: %v = VADD_VV ... %s = VREDSUM w/ %v as scalar source %dead = VADD_VV %v, %v w/ VL=0
> 
> The last instruction is dead - it can be folded to it's passthru. (In practice, it probably will have been folded, but it's possible something could slip through to here.) However, when scaning the users of %v, we will decide that the correct VL for %v is 0 (or a register which might be zero), and reduce it below the minimum VL=1 required by the reduction.
> 
> To fix this, I believe you need to treat the CommonVL for the scalar operand case as being VL=1. You could also track a non-zero state instead.
> 
> Other than that, looks good to me. Once you've fixed this issue, happy to approve.

Should we just remove this code for now

```
    // Instructions like reductions may use a vector register as a scalar
    // register. In this case, we should treat it like a scalar register which
    // does not impact the decision on whether to optimize VL.
    if (isVectorOpUsedAsScalarOp(UserOp)) {
      [[maybe_unused]] Register R = UserOp.getReg();
      [[maybe_unused]] const TargetRegisterClass *RC = MRI->getRegClass(R);
      assert(RISCV::VRRegClass.hasSubClassEq(RC) &&
             "Expect LMUL 1 register class for vector as scalar operands!");
      LLVM_DEBUG(dbgs() << "    Use this operand as a scalar operand\n");
      continue;
    }
```

**Comment by michaelmaitland** [comment]
> > Should we just remove this code for now
> 
> This would be fine by me. Incrementalism is good. :)

I've removed it in https://github.com/llvm/llvm-project/pull/120291.

FWIW, I don't think what you're concerned about can happen with or without https://github.com/llvm/llvm-project/pull/120291 merged since all the instructions that isVectorOpUsedAsScalarOp deal with return `OperandInfo(Unknown)` and won't lead to any (incorrect) optimization.

**Comment by topperc** [comment]
> > > Should we just remove this code for now
> > 
> > 
> > This would be fine by me. Incrementalism is good. :)
> 
> I've removed it in #120291.
> 
> FWIW, I don't think what you're concerned about can happen with or without #120291 merged since all the instructions that isVectorOpUsedAsScalarOp deal with return `OperandInfo(Unknown)` and won't lead to any (incorrect) optimization.

The code that was there said that we could ignore the reduction and not call getOperandInfo on it. It effectively said that a scalar operand doesn't depend on any elements from the producer. This is incorrect, it demands exactly 1 element. With that code in place only the VL of the other consumers was used. If they used less than 1 element then the 1 element that scalar op needs wouldn't be valid.

**Comment by michaelmaitland** [comment]
> > > > Should we just remove this code for now
> > > 
> > > 
> > > This would be fine by me. Incrementalism is good. :)
> > 
> > 
> > I've removed it in #120291.
> > FWIW, I don't think what you're concerned about can happen with or without #120291 merged since all the instructions that isVectorOpUsedAsScalarOp deal with return `OperandInfo(Unknown)` and won't lead to any (incorrect) optimization.
> 
> The code that was there said that we could ignore the reduction and not call getOperandInfo on it. So it doesn't matter that the reduction is missing from getOperandInfo. The code effectively said that a scalar operand doesn't depend on any elements from the producer. This is incorrect, it demands exactly 1 element. With that code in place only the VL of the other consumers was used. If they used less than 1 element then the 1 element that scalar op needs wouldn't be valid.

Yes, my bad.

**Review by preames — APPROVED** [review]
LGTM

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-changes-length.ll [hunk 1]
```diff
@@ -97,8 +97,9 @@ define <4 x i32> @v4i32_v8i32(<8 x i32>) {
 define <4 x i32> @v4i32_v16i32(<16 x i32>) {
 ; RV32-LABEL: v4i32_v16i32:
 ; RV32:       # %bb.0:
-; RV32-NEXT:    vsetivli zero, 8, e16, m1, ta, ma
+; RV32-NEXT:    vsetivli zero, 2, e16, m1, ta, ma
 ; RV32-NEXT:    vmv.v.i v12, 1
+; RV32-NEXT:    vsetivli zero, 8, e16, m1, ta, ma
 ; RV32-NEXT:    vmv.v.i v14, 6
```
Comment (lukel97) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-changes-length.ll:103]: Do we know why only one of the vmv.v.is had their VL reduced here?

Comment (michaelmaitland) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-shuffle-changes-length.ll:103]: Take a look at the MIR: https://godbolt.org/z/xrvG13qx6

You can see that `%4` is used as a tied operand.

https://github.com/llvm/llvm-project/blob/644643a4ee931ce24c27702851a93160cf0f1248/llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp#L917

**Pull request merged**